### DISCUSSION TITLE
Use in Linux

### DISCUSSION THREAD

**Discussion by void-spark** [dioannidis/usbasp#15] [General]
Any idea if this firmware works in Linux?
Would it also need a custom avrdude that supports composite devices or something similar?
Any special options that would be needed, like a specific -P ?
If not, any suggestions for a branch which does work under Linux? :)

**Comment by dioannidis** [comment]
Hi @void-spark,
<br/><br/><br/>
> Any idea if this firmware works in Linux?

  yes it works . At least in distros who support USB 2.0 BOS descriptors . 

  i.e. see Debian Buster and up ( running in VM ): 
 
 ![USBasp-dmesg](https://user-images.githubusercontent.com/15191633/197214873-513dce74-6a48-4452-ab76-8ef835aa08c3.PNG)
<br/><br/><br/>
> Would it also need a custom avrdude that supports composite devices or something similar?

AFAIK, avrdude in linux uses libusb which supports composite devices, so no there is no need . 

i.e. Arduino IDE 1.8.3 Debian Bullseye uploading the ASCII Table sketch using USBasp as the progammer : 

![USBasp-Arduino](https://user-images.githubusercontent.com/15191633/197215697-218fa0e3-791f-435f-9a22-9405f9c2387f.PNG)
<br/><br/><br/>
> Any special options that would be needed, like a specific -P ?

Nope, just the usual -c usbasp .

Also there is a linux x86_64 version of the USBaspHIDUART console utility which at least runs in Debian : 

i.e.

![USBasp-list](https://user-images.githubusercontent.com/15191633/197216586-849e43b5-a3a6-437e-a5bd-753c5427dd11.PNG)




and an example of reading the output of the ASCII Table sketch using USBaspHIDUART :

![USBasp-recieve](https://user-images.githubusercontent.com/15191633/197216838-3340cc36-f4ff-4d6e-827e-0e47fd2817b4.PNG)

regards,

**Comment by void-spark** [comment]
Yep, been able to test now, it works! (at least the programming part).
I made it a bit more difficult for myself by running avrdude from a docker container, but --privileged helped with that :)

I guess the uart part is not a standard hid 'profile(?)', so it always needs a custom program that uses usblib to talk with it?
Just out of interest, I have so many usb to serial converters, so don't need it much :)

**Reply by dioannidis** [reply]
Hi,

> I guess the uart part is not a standard hid 'profile(?)', so it always needs a custom program that uses usblib to talk with it? 

A serial HID profile doesn't exists AFAIK ... So yes, you need a software that implements the [uart HID protocol](https://github.com/dioannidis/usbasp#uart-hid-protocol) to talk to the USBasp uart which talks to the target mcu uart. 

The [USBaspHIDUART console utility](https://github.com/dioannidis/usbasp#usbasphiduart-utility) has implement this protocol and can be used for simple stuff .

regards,
